pywemo/ssdp.py

```diff
@@ -336,16 +366,137 @@ def scan(
+
+    Wemo devices are informed of the pywemo virtual Wemo device in two ways:
+
+    1. Wemo devices periodically send UPnP M-SEARCH discovery requests for the
+    to locate other devices on the network. DiscoveryResponder responds to these
+    requests with the URL for the virtual device.
+
+    2. A UPnP NOTIFY message is periodically multicasted by DiscoveryResponder
+    to inform Wemo devices on the network of the URL for the virtual device.
+    """
+
+    def __init__(self, callback_port: int):
+        """Create a server that will respond to WeMo discovery requests.
+
+        Args:
+            callback_port: The port for the SubscriptionRegistry HTTP server.
+        """
+        self.callback_port = callback_port
+        self._thread = None
+        self._exit = threading.Event()
+        self._thread_exception = None
+        self._notify_enabled = True  # For tests.
+
+    def send_notify(self) -> None:
+        """Send a UPnP NOTIFY message containing the virtual device URL."""
+        ssdp_target = (MULTICAST_GROUP, MULTICAST_PORT)
+        for addr in interface_addresses():  # Send on all interfaces.
+            callback_addr = (addr, self.callback_port)
+            sock = socket.socket(socket.AF_INET, socket.SOCK_DGRAM)
+            try:
+                sock.bind((addr, 0))
+                sock.sendto(
+                    (SSDP_NOTIFY % callback_addr).encode("UTF-8"), ssdp_target
+                )
+            except socket.error:
+                pass
+            finally:
+                sock.close()
+
+    def respond_to_discovery(self) -> None:
+        """Respond to a WeMo discovery request with the virtual device URL."""
+        sock = socket.socket(socket.AF_INET, socket.SOCK_DGRAM)
+        try:
+            sock.setsockopt(socket.SOL_SOCKET, socket.SO_REUSEADDR, 1)
+
+            # Join the multicast group on all interfaces.
+            group = socket.inet_aton(MULTICAST_GROUP)
+            for addr in interface_addresses():
+                try:
+                    local = socket.inet_aton(addr)
+                    sock.setsockopt(
+                        socket.IPPROTO_IP,
+                        socket.IP_ADD_MEMBERSHIP,
+                        group + local,
+                    )
+                except socket.error:
+                    pass
+
+            sock.bind((MULTICAST_GROUP, MULTICAST_PORT))
+
+            next_notify = datetime.min
+            while not self._exit.is_set():
+                # Periodically send NOTIFY messages.
+                now = datetime.now()
+                if now > next_notify and self._notify_enabled:
+                    next_notify = now + timedelta(minutes=2)
+                    self.send_notify()
+
+                # Check for new discovery requests.
+                if not select.select([sock], [], [], 1)[0]:
+                    continue  # Timeout, no data. Loop again and check for exit.
+                msg, addr = sock.recvfrom(1024)
```

Comment:
Is 1024 just the right number for the M-SEARCH protocol?

Reply:
I used the same value (1024) as is used for discovery. 1500 might be more appropriate for both, but either value will work. None of the M-SEARCH messages I have seen have been that large.